cannot include dependencies not required by bundle (knex pg).

Hi,

I'm using knex and bookshelf with pg-library, the problem is that the pg library does not get bundled in to node_modules directory because it is not required in the files that get bundled (but is required by knex on runtime). This causes AWS Lambda to throw following errors.

2017-09-08T07:28:40.829Z	52dbac4f-9467-11e7-af3a-fdbf9a737897	Knex: run
$ npm install pg --save
Error: Cannot find module 'pg'
at Function.Module._resolveFilename (module.js:469:15)
at Function.Module._load (module.js:417:25)
at Module.require (module.js:497:17)
at require (internal/module.js:20:19)
at Client_PG._driver (/var/task/node_modules/knex/lib/dialects/postgres/index.js:91:12)
at Client_PG.initializeDriver (/var/task/node_modules/knex/lib/client.js:219:26)
at Client_PG.Client (/var/task/node_modules/knex/lib/client.js:113:10)
at new Client_PG (/var/task/node_modules/knex/lib/dialects/postgres/index.js:56:20)
at Knex (/var/task/node_modules/knex/lib/index.js:60:34)
at Object.module.exports.dev.elasticsearch (/var/task/index.js:102:35)
at __webpack_require__ (/var/task/index.js:20:30)
at Object.<anonymous> (/var/task/index.js:79:17)
at __webpack_require__ (/var/task/index.js:20:30)
at Object.<anonymous> (/var/task/index.js:270:12)
at __webpack_require__ (/var/task/index.js:20:30)
at Object.<anonymous> (/var/task/index.js:217:15)

Probably the cause is that this is not direct depency but resolved one during runtime which is then not resolved by webpack during bundling. I would appreciate a lot if this could work, although i could also probably just:

 var dump = require("pg"); 

in order to get it bundled for now

--- a/README.md
+++ b/README.md
@@ -192,6 +192,24 @@ custom:
 ```
 > Note that only relative path is supported at the moment.
 
+
+Sometimes it might happen that you use dynamic requires in your code, i.e. you
+require modules that are only known at runtime. Webpack is not able to detect
+such externals and the compiled package will miss the needed dependencies.
+In such cases you can force the plugin to include certain modules by setting
+them in the `forceInclude` array property. However the module must appear in
+your service's production dependencies in `package.json`.
+
+```yaml
+# serverless.yml
+custom:
+  webpackIncludeModules:
+    forceInclude:
+      - module1
+      - module2
+```
+
+
 You can find an example setups in the [`examples`][link-examples] folder.
 
 #### Service level packaging

--- a/lib/packExternalModules.js
+++ b/lib/packExternalModules.js
@@ -110,6 +110,7 @@ module.exports = {
       return BbPromise.resolve(stats);
     }
 
+    const packageForceIncludes = _.get(includes, 'forceInclude', []);
     const packagePath = includes.packagePath || './package.json';
     const packageJsonPath = path.join(process.cwd(), packagePath);
 
@@ -158,7 +159,10 @@ module.exports = {
 
       // (1) Generate dependency composition
       const compositeModules = _.uniq(_.flatMap(stats.stats, compileStats => {
-        const externalModules = getExternalModules.call(this, compileStats);
+        const externalModules = _.concat(
+          getExternalModules.call(this, compileStats),
+          _.map(packageForceIncludes, whitelistedPackage => ({ external: whitelistedPackage }))
+        );
         return getProdModules.call(this, externalModules, packagePath, dependencyGraph);
       }));
 
@@ -196,7 +200,11 @@ module.exports = {
         const modulePackage = {
           dependencies: {}
         };
-        const prodModules = getProdModules.call(this, getExternalModules.call(this, compileStats), packagePath, dependencyGraph);
+        const prodModules = getProdModules.call(this,
+          _.concat(
+            getExternalModules.call(this, compileStats),
+            _.map(packageForceIncludes, whitelistedPackage => ({ external: whitelistedPackage }))
+          ), packagePath, dependencyGraph);
         _.forEach(prodModules, prodModule => {
           const splitModule = _.split(prodModule, '@');
           // If we have a scoped module we have to re-add the @

--- a/tests/mocks/package.mock.json
+++ b/tests/mocks/package.mock.json
@@ -8,7 +8,8 @@
     "npm-programmatic": "0.0.5",
     "uuid": "^5.4.1",
     "ts-node": "^3.2.0",
-    "@scoped/vendor": "1.0.0"
+    "@scoped/vendor": "1.0.0",
+    "pg": "^4.3.5"
   },
   "devDependencies": {
     "babel-eslint": "^7.2.3",

--- a/tests/packExternalModules.test.js
+++ b/tests/packExternalModules.test.js
@@ -349,5 +349,107 @@ describe('packExternalModules', () => {
         expect(childProcessMock.exec).to.have.been.calledOnce,
       ]));
     });
+
+    it('should install external modules when forced', () => {
+      const expectedPackageJSON = {
+        dependencies: {
+          '@scoped/vendor': '1.0.0',
+          uuid: '^5.4.1',
+          bluebird: '^3.4.0',
+          pg: '^4.3.5'
+        }
+      };
+      serverless.service.custom = {
+        webpackIncludeModules: {
+          forceInclude: ['pg']
+        }
+      };
+      module.webpackOutputPath = 'outputPath';
+      npmMock.install.returns(BbPromise.resolve());
+      fsExtraMock.copy.yields();
+      childProcessMock.exec.onFirstCall().yields(null, '{}', '');
+      childProcessMock.exec.onSecondCall().yields();
+      return expect(module.packExternalModules(stats)).to.be.fulfilled
+      .then(() => BbPromise.all([
+        // npm install should have been called with all externals from the package mock
+        expect(npmMock.install).to.have.been.calledOnce,
+        expect(npmMock.install).to.have.been.calledWithExactly([
+          '@scoped/vendor@1.0.0',
+          'uuid@^5.4.1',
+          'bluebird@^3.4.0',
+          'pg@^4.3.5'
+        ],
+        {
+          cwd: path.join('outputPath', 'dependencies'),
+          maxBuffer: 204800,
+          save: true
+        }),
+        // The module package JSON and the composite one should have been stored
+        expect(writeFileSyncStub).to.have.been.calledTwice,
+        expect(writeFileSyncStub.firstCall.args[1]).to.equal('{}'),
+        expect(writeFileSyncStub.secondCall.args[1]).to.equal(JSON.stringify(expectedPackageJSON, null, 2)),
+        // The modules should have been copied
+        expect(fsExtraMock.copy).to.have.been.calledOnce,
+        // npm ls and npm prune should have been called
+        expect(childProcessMock.exec).to.have.been.calledTwice,
+        expect(childProcessMock.exec.firstCall).to.have.been.calledWith(
+          'npm ls -prod -json -depth=1'
+        ),
+        expect(childProcessMock.exec.secondCall).to.have.been.calledWith(
+          'npm prune'
+        )
+      ]));
+    });
+
+    it('should add forced external modules without version when not in production dependencies', () => {
+      const expectedPackageJSON = {
+        dependencies: {
+          '@scoped/vendor': '1.0.0',
+          uuid: '^5.4.1',
+          bluebird: '^3.4.0',
+          'not-in-prod-deps': ''
+        }
+      };
+      serverless.service.custom = {
+        webpackIncludeModules: {
+          forceInclude: ['not-in-prod-deps']
+        }
+      };
+      module.webpackOutputPath = 'outputPath';
+      npmMock.install.returns(BbPromise.resolve());
+      fsExtraMock.copy.yields();
+      childProcessMock.exec.onFirstCall().yields(null, '{}', '');
+      childProcessMock.exec.onSecondCall().yields();
+      return expect(module.packExternalModules(stats)).to.be.fulfilled
+      .then(() => BbPromise.all([
+        // npm install should have been called with all externals from the package mock
+        expect(npmMock.install).to.have.been.calledOnce,
+        expect(npmMock.install).to.have.been.calledWithExactly([
+          '@scoped/vendor@1.0.0',
+          'uuid@^5.4.1',
+          'bluebird@^3.4.0',
+          'not-in-prod-deps'
+        ],
+        {
+          cwd: path.join('outputPath', 'dependencies'),
+          maxBuffer: 204800,
+          save: true
+        }),
+        // The module package JSON and the composite one should have been stored
+        expect(writeFileSyncStub).to.have.been.calledTwice,
+        expect(writeFileSyncStub.firstCall.args[1]).to.equal('{}'),
+        expect(writeFileSyncStub.secondCall.args[1]).to.equal(JSON.stringify(expectedPackageJSON, null, 2)),
+        // The modules should have been copied
+        expect(fsExtraMock.copy).to.have.been.calledOnce,
+        // npm ls and npm prune should have been called
+        expect(childProcessMock.exec).to.have.been.calledTwice,
+        expect(childProcessMock.exec.firstCall).to.have.been.calledWith(
+          'npm ls -prod -json -depth=1'
+        ),
+        expect(childProcessMock.exec.secondCall).to.have.been.calledWith(
+          'npm prune'
+        )
+      ]));
+    });
   });
 });

